bl-recent-files-pipemenu.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH BL-RECENT-FILES-PIPEMENU "1" "July 2025" "bl-recent-files-pipemenu 13.0-1" "User Commands"
.SH NAME
bl-recent-files-pipemenu \- a script provided by bunsen-pipemenus
.SH SYNOPSIS
.B bl-recent-files-pipemenu
[\fI\,-h\/\fR] [\fI\,-f XBEL\/\fR] [\fI\,-l NUM\/\fR] [\fI\,-o CMD\/\fR] [\fI\,-r\/\fR]
.SH DESCRIPTION
Where: \fB\-h\fR       (help) Show this message and exit
.TP
\fB\-f\fR
(file) Specify the path to a 'recently\-used.xbel' file, omit in order to auto\-detect
.TP
\fB\-l\fR
(limit) Include up to NUM items in the menu
.TP
\fB\-o\fR
(open\-command) Open files using CMD, by default files will be opened in their original application
.TP
\fB\-r\fR
(reverse) Put the most\-recently used items at the top of the menu
.SH AUTHOR
Written by the BunsenLabs team.
.SH "REPORTING BUGS"
Please report bugs at:

https://github.com/BunsenLabs/bunsen-pipemenus/issues
.SH "SEE ALSO"
The BunsenLabs forums may be able to answer your questions:

https://forums.bunsenlabs.org


debian/changelog
bunsen-pipemenus (13.1-1) carbon; urgency=medium

  * bl-recent-files-pipemenu: remove long options which are
     no longer supported by lua-posix 34.0.
  * Update documentation.
  * genman.sh: update to Version 20250802 which supports passing
     extra arguments to help2man.
  * Remove tint2-pipemenu from autogenerated menus.
     (The pipemenu is still shipped in this package.)

 -- John Crawley <john@bunsenlabs.org>  Sat, 02 Aug 2025 13:49:45 +0900

bunsen-pipemenus (13.0-1) carbon; urgency=medium

  * Migrate for Debian Bookworm
    - debian/control: Bump Standards-Version to 4.6.2 - no changes required.
  * Add notes about bl-kb, bl-xbk and bl-jgpop.

 -- John Crawley <john@bunsenlabs.org>  Tue, 25 Jun 2024 15:08:32 +0900

bunsen-pipemenus (12.2.3-1) boron; urgency=medium

  * bl-sshconfig-pipemenu: tweak prompt to close terminal.

 -- John Crawley <john@bunsenlabs.org>  Sun, 07 Apr 2024 16:01:17 +0900

bunsen-pipemenus (12.2.2-1) boron; urgency=medium

  * Fix autostart path in Dropbox pipemenu.

 -- John Crawley <john@bunsenlabs.org>  Fri, 26 Jan 2024 18:13:35 +0900

bunsen-pipemenus (12.2.1-2) boron; urgency=medium

  * Update copyright year.

 -- John Crawley <john@bunsenlabs.org>  Thu, 11 Jan 2024 14:52:16 +0900

bunsen-pipemenus (12.2.1-1) boron; urgency=medium

  * bl-install-pipemenu: fix package name of openshot-qt from openshot.

 -- John Crawley <john@bunsenlabs.org>  Thu, 11 Jan 2024 14:47:44 +0900

bunsen-pipemenus (12.2-1) boron; urgency=medium

  * Move bl-kb and bl-xbk here from bunsen-utilities.
     They are used by bl-kb-pipemenu.
  * bl-kb: Output destination of openbox keybinds:
     GoToDesktop, SendToDesktop and DirectionalCycleWindows
  * bl-kb: do not print to stdout - output goes directly to a file.
  * bl-kb-pipemenu: rename option --keybinds to --kbmenu.
  * bl-kb-pipemenu: rearrange items when outputting keybinds as menu,
     with openbox at the top.
  * bl-kb-pipemenu: comment out "Backup and Edit..." items from pipemenu.
     They will be in the parent menu.
  * bl-compositor: pipemenu: put config file edit items at the top.

 -- John Crawley <john@bunsenlabs.org>  Fri, 06 Oct 2023 16:19:51 +0900

bunsen-pipemenus (12.1-1) boron; urgency=medium

  * bl-install-pipemenu: add Dropbox item under "Utilities".

 -- John Crawley <john@bunsenlabs.org>  Tue, 11 Jul 2023 09:36:35 +0900

bunsen-pipemenus (12.0-1) boron; urgency=medium

  * Migrate for Debian Bookworm
    - debian/control: Bump Standards-Version to 4.6.1.0 - no changes required.
    - debian/control: Declare debhelper compatibility 13.
    - debian/control: Add Rules-Requires-Root: no
    - debian/control: Add previously missing Build-Depends: python3-lxml
    - debian/copyright: update year, use https Format url.

 -- John Crawley <john@bunsenlabs.org>  Sun, 22 Jan 2023 14:21:26 +0900

bunsen-pipemenus (11.3.1-1) beryllium; urgency=medium

  * Fix helpfile paths in conky & tint2 pipemenus.
  * bl-install-pipemenu: mdadm: add mention of RAID.

 -- John Crawley <john@bunsenlabs.org>  Sat, 14 May 2022 15:33:27 +0900

bunsen-pipemenus (11.3-1) beryllium; urgency=medium

  * Add bl-jgpop, utility for calling standalone instance of a pipemenu.
  * Add desktop files for compositor, conky and tint2 pipemenus.
  * bl-tint2-pipemenu: add item to restore default tint2 session.
  * bl-install-pipemenu:
     - replace gnome-mpv with celluloid
     - replace gtk-recordmydesktop with simplescreenrecorder
     - add "Utilities" section with hw-probe, mc, mdadm and gnome-calculator
  * Remove bl-help-pipemenu, superseded by an include from bunsen-configs.

 -- John Crawley <john@bunsenlabs.org>  Wed, 12 Jan 2022 10:21:40 +0900

bunsen-pipemenus (11.2-1) beryllium; urgency=medium

  * bl-conky-pipemenu, bl-tint2-pipemenu:
     adjust to match updated scripts in bunsen-utilities 11.1-1.
  * bl-conky-pipemenu: add item to set conky startup session.
  * bl-compositor: use bl-restart instead of bl-tint2restart.

 -- John Crawley <john@bunsenlabs.org>  Thu, 19 Aug 2021 15:41:18 +0900

bunsen-pipemenus (11.1-1) beryllium; urgency=medium

  * bl-compositor: switch from compton to picom.
     (bl-compositor is still usable if compton is installed instead of picom.)

 -- John Crawley <john@bunsenlabs.org>  Fri, 21 May 2021 15:40:05 +0900

bunsen-pipemenus (11.0-1) beryllium; urgency=medium

  * Migrate for Debian Bullseye.
    + debian/control: Bump Standards-Version to 4.5.1 - no changes required.
    + debian/control: Declare debhelper compatibility 12.
    + Remove debian/compat, now unnecessary.
    + debian/copyright: update year.
  * bl-install-pipemenu: replace libreoffice-gtk2 with libreoffice-gtk3
  * Remove unused pipemenus.

 -- John Crawley <john@bunsenlabs.org>  Fri, 12 Feb 2021 17:20:36 +0900

bunsen-pipemenus (10.3-1) lithium; urgency=medium

  * bl-dropbox-pipemenu: Update download url for dropbox.py.
  [ jmechnich ]
  * bl-sshconfig-pipemenu: Port to python3.
  [ Jens John ]
  * bl-sshconfig-pipemenu: Improve code.

 -- John Crawley <john@bunsenlabs.org>  Wed, 13 Jan 2021 17:59:45 +0900

bunsen-pipemenus (10.2.5-1) lithium; urgency=medium

  * tint2-pipemenu: Remove tint2conf entry.

 -- John Crawley <john@bunsenlabs.org>  Thu, 16 Jul 2020 14:33:27 +0900

bunsen-pipemenus (10.2.4-1) lithium; urgency=medium

  * bl-install-pipemenu: add "Web Browsers" section.

 -- John Crawley <john@bunsenlabs.org>  Sat, 11 Jul 2020 18:18:11 +0900

bunsen-pipemenus (10.2.3-1) lithium; urgency=medium

  * bl-{conky,tint2}-pipemenu: Relabel "Chooser" to "Manager".
  * debian/control: add lua and python Build-Depends for man pages.

 -- John Crawley <john@bunsenlabs.org>  Sat, 16 May 2020 14:21:44 +0900

bunsen-pipemenus (10.2.2-1) lithium; urgency=medium

  * bl-beepmein-pipemenu: add --standalone option for launchers.
  * bl-dropbox-pipemenu: "dropbox": make symlink to "dropbox.py" if existing.

 -- John Crawley <john@bunsenlabs.org>  Mon, 20 Apr 2020 11:45:51 +0900

bunsen-pipemenus (10.2.1-1) lithium; urgency=medium


bl-recent-files-pipemenu.lua
#!/usr/bin/env lua5.2

local posix = require 'posix'
local url = require 'socket.url'
local xml = require 'pl.xml'
-- compatibility with Lua 5.1/LuaJIT
if _VERSION == "Lua 5.1" then
  table.unpack = unpack
end

local M = xml.new("openbox_pipe_menu")

local function swap(a, b)
  return b, a
end

local function quote(s)
  return string.format("%q", s)
end

local function printf(f, ...)
  io.stderr:write(string.format(f, ...) .. '\n')
end

local function decode_url(s)
  local s = s

  s = s:gsub("_", "__")

  if s:find("%", 1, true) then
    s = url.unescape(s)
  end

  return s
end

local function mk_item_tag(label, actions)
  return xml.elem("item", { label = label, table.unpack(actions)})
end

local function mk_action_tag(name, command_tag)
  return xml.elem("action", { name = name, command_tag })
end

local function mk_command_tag(cmd, args)
  return xml.elem("command"):text(quote(cmd) .. ' ' .. quote(args))
end

local function add_executable_item(m, label, cmd, args)
  m:add_direct_child(
    mk_item_tag(label, {
      mk_action_tag("Execute", mk_command_tag(cmd, args))
    }))
end

local function add_items_from_xbel(m, path, reverse_output, cnt, override_cmd)
  if not posix.access(path) then return false end

  local function clean_executable_name(s)
    return s:sub(2, -5)
  end

  local x = xml.parse(path, true)
  if not x then return false end

  local buf, c = {}, 0

  for bookmark in x:childtags() do
    local file = bookmark:get_attribs().href
    local apps = bookmark:get_elements_with_name("bookmark:application")
    local prop = apps[#apps]:get_attribs()
    local i = mk_item_tag(decode_url(posix.basename(file)), {
        mk_action_tag("Execute", mk_command_tag(override_cmd or clean_executable_name(prop.exec), file))
      })
    if reverse_output then
      table.insert(buf, 1, i)
    else
      table.insert(buf, i)
    end
  end

  for i,v in ipairs(buf) do
    if cnt and i > cnt then break end
    m:add_direct_child(v)
  end
end

local function add_remove_file_item(m, path)
  m:add_direct_child(xml.elem("separator"))
  m:add_direct_child(
    mk_item_tag("Clear recent files",
      mk_action_tag("Execute", mk_command_tag("/bin/bash", "-c '> " .. path .. "'")))
    )
end

local function add_is_empty_message(m)
  m:add_direct_child(
    mk_item_tag("No recently used files.", {})
  )
end

local function print_menu(m)
  print(xml.tostring(m, "", "  "))
end

local function detect_xbel_path()
  local hp = os.getenv("HOME")
  local xh = os.getenv("XDG_DATA_HOME")
  local fn = "recently-used.xbel"
  local function check_path(t)
    for _,p in ipairs(t) do
      if posix.access(p) then return p end
    end
    return nil
  end
  return check_path{
    string.format("%s/%s", xf, fn),
    string.format("%s/.local/share/%s", hp, fn),
    string.format("%s/.%s", hp, fn)
  }
end

local function main()
  local xbel_path, reverse, cnt, override = detect_xbel_path(), false, false, false
  for o, optarg, opterr in posix.getopt(arg, "hf:l:o:r", {
    { "limit",    "required", 'l' },
    { "help",     "none",     'h' },
    { "file",     "required", 'f' },
    { "reverse",  "none",     'r' },
    { "open-cmd", "required", 'o' }}) do
    if o == '?' then
      printf("Invalid option or missing argument: %s", arg[opterr-1])
      return 1
    elseif o == 'h' then
      printf([[Usage: %s [-h|--help] [-f|--file XBEL] [-l|--limit NUM] [-o|--open-cmd CMD] [-r|--reverse]
Where: -h, --help       Show this message and exit.
       -f, --file       XBEL. Path to a 'recently-used.xbel' file.
                        Omit in order to attempt to auto-detect the path.
       -l, --limit      NUM. Include up to NUM items in the menu.
       -o, --open-cmd   CMD. Open files using CMD. By default, files will be
                        opened in the application they were being used with.
       -r, --reverse    Put the most-recently used items at the top of the menu.]], arg[0])
      return 0
    elseif o == 'f' then
      xbel_path = optarg
      if not posix.access(xbel_path) then
        printf("Warning: User-specified XBEL file doesn't exist.", xbel_path)
      end
    elseif o == 'l' then
      cnt = tonumber(optarg)
    elseif o == 'o' then
      override = optarg
    elseif o == 'r' then
      reverse = true
    end
  end

  add_items_from_xbel(M, xbel_path, reverse, cnt, override)

  if #M == 0 then
    add_is_empty_message(M)
  else
    add_remove_file_item(M, xbel_path)
  end

  print_menu(M)

  return 0
end

os.exit(main())
